### PR TITLE
In-memory AST for generated C source code

### PR DESCRIPTION
**Pull request by navjotk**
This branch addresses Issue [#57](https://github.com/opesci/opesci-fd/issues/57)
The C source code generation has been updated to use the [Cgen](https://github.com/inducer/cgen) python module. 
However this is still not a complete in-memory AST as the syntax tree is only maintained at the code-block level and is finally converted to a string and added to a template by [grid.py](https://github.com/opesci/opesci-fd/blob/cgen_ast/opesci/grid.py#L66). I would like to get some feedback on the changes so far before I go ahead and change this last function. 
Even though this is not a complete in-memory AST, this pull-request already achieves two of the objectives listed in Issue [#57](https://github.com/opesci/opesci-fd/issues/57), i.e. code readability and maintainability have improved and so has the pretty-print of the generated C code.

### CODE REVIEW

File: opesci/cgen_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from cgen import *
+
+
+class IfDef(Module):
```
Comment (ctjacobs) [inline opesci/cgen_wrapper.py]: """ docstring please """

File: opesci/cgen_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self, condition, iflines, elselines):
+        ifdef_line = Line('#ifdef %s' % condition)
+        else_line = Line('#else')
+        endif_line = Line('#endif')
+        lines = [ifdef_line]+iflines+[else_line]+elselines+[endif_line]
+        super(IfDef, self).__init__(lines)
+
+
+class InlineInitializer(Initializer):
+    def generate(self):
```
Comment (ctjacobs) [inline opesci/cgen_wrapper.py]: """ docstring please """

File: opesci/cgen_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        result = super(InlineInitializer, self).generate()
+        for v in result:
+            if v.endswith(';'):
+                yield v[:-1]
+            else:
+                yield v
+
+
+def replace_in_code(code, str_from, str_to):
```
Comment (ctjacobs) [inline opesci/cgen_wrapper.py]: """ docstring please """

**Comment by mlange05** [comment]
First of all, this is an excellent contribution; many thanks @navjotk. Our generated code now looks human readable! I only have a few minor comments:
* In some places we ask the compiler for the right pragma and insert a `cgen.Statement` and in some we just insert `cgen.Pragma("ivdep")`. We should really be consistent and use `cgen.Pragma(self.compiler._ivdep)` and drop the `#pragam` in compilation.py. Please note that this was most likely broken before, but now is still a good time to fix it.
* Looking at a before-and-after diff it seems that a few `const` declarations either get jumbled or dropped.
* Since we dropped the template generation with this there is a whole lot of further clean up we can do by removing the template .txt files, the dependency on `Mako` and `mpmath` (I think). We can leave this for another branch/PR though.

Apart from these minor comments I'm very happy for this to go in. Thanks again!

**Comment by navjotk** [comment]
Thanks for the comments @mlange05 . The last [commit](https://github.com/opesci/opesci-fd/commit/bce4a2c096cc3b4be5db001a36fa72982a812d84) addresses the first of your concerns.

About your second concern, as we discussed earlier, the amount of effort required to switch the order of const and type keywords in the cgen generated code doesn't seem like its worth the effort since the code would just compile fine. I agree that it litters our diff output right now. 
I did another diff to double check any missing constants as you mention. Unfortunately I couldn't find the missing constants you mention. The diff output I saw is posted [here](https://gist.github.com/navjotk/195da9717dd3d07226fa).   

I believe the last of your concerns is best addressed, as you mention, in its own branch/PR. I am glad to take that up too, depending on priority.

**Comment by mlange05** [comment]
Ok, confirmed the `const` juggling is fine. Apart from a small typo and a PEP8 violation that crept in with the last commit I think this is good to go, as soon as travis is green. Many thanks again.

**Comment by navjotk** [comment]
Fixed the typo that was causing Travis to fail previously. This time the travis build failed at downloading pluto, something I believe I haven't touched. Can someone help me figure this out please?
Also, for curiosity, @mlange05 , was the PEP8 violation you mentioned, the print statement?

**Comment by mlange05** [comment]
Ok, it seems there is a problem with the sourceforge server where we get pluto from. If this isn't fixed soon, I'll disable the pluto tests, so that we can merge this. In that case you might need to rebase your branch over the recent fixes.

**Comment by navjotk** [comment]
Rebased the recent fixes. Also travis is green now.

**Comment by mlange05** [comment]
Great, many thanks again.

**Pull request merged**